agent/csi/plugin.go

```diff
@@ -33,12 +50,133 @@ type NodePlugin struct {
 
 	// NodeClient is the node service client
 	NodeClient csi.NodeClient
+
+	// VolumeMap is the map from volume ID to Volume. Will place a volume once it is staged,
+	// remove it from the map for unstage.
+	// TODO: Make this map persistent if the swarm node goes down
+	VolumeMap map[string]*volumePublishStatus
+
+	// Lock for VolumeMap
+	Lock sync.RWMutex
+}
+
+const TargetPath string = "/var/lib/docker/%s"
+
+func NewNodePlugin(name string) *NodePlugin {
+	return &NodePlugin{
+		Name:      name,
+		VolumeMap: make(map[string]*volumePublishStatus),
+	}
+}
+
+//// Get returns a volume published path for the provided volume ID.  If the volume doesn't exist, returns empty string.
+func (np *NodePlugin) GetPublisedPath(volumeID string) (string, error) {
+	np.Lock.RLock()
+	defer np.Lock.RUnlock()
+	if volInfo, ok := np.VolumeMap[volumeID]; ok {
+		if volInfo.isPublished {
+			return volInfo.targetPath, nil
+		}
+		return volInfo.targetPath, fmt.Errorf("volume %s is not published", volumeID)
+	}
+	return "", nil
 }
 
 func (np *NodePlugin) NodeGetInfo(ctx context.Context) (*api.NodeCSIInfo, error) {
+	np.Lock.RLock()
+	defer np.Lock.RUnlock()
 	resp := &csi.NodeGetInfoResponse{
 		NodeId: np.NodeID,
 	}
 
 	return makeNodeInfo(resp), nil
 }
+
+func (np *NodePlugin) NodeStageVolume(ctx context.Context, req *api.VolumeAssignment) error {
+
+	volID := req.VolumeID
+	stagingTarget := fmt.Sprintf(TargetPath, volID)
+
+	// Check arguments
+	if len(volID) == 0 {
+		return status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+
+	np.Lock.Lock()
+	defer np.Lock.Unlock()
+
+	v := &volumePublishStatus{
+		targetPath: stagingTarget,
+	}
+
+	np.VolumeMap[volID] = v
+
+	return nil
+}
+
+func (np *NodePlugin) NodeUnstageVolume(ctx context.Context, req *api.VolumeAssignment) error {
+
+	volID := req.VolumeID
+
+	// Check arguments
+	if len(volID) == 0 {
+		return status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+
+	np.Lock.Lock()
+	defer np.Lock.Unlock()
+	if v, ok := np.VolumeMap[volID]; ok {
+		if v.isPublished {
+			return status.Errorf(codes.InvalidArgument, "VolumeID %s is not unpublished", volID)
+		}
+		delete(np.VolumeMap, volID)
+		return nil
+	}
+
+	return status.Errorf(codes.NotFound, "VolumeID %s is not staged", volID)
+}
+
+func (np *NodePlugin) NodePublishVolume(ctx context.Context, req *api.VolumeAssignment) error {
+
+	volID := req.VolumeID
+
+	// Check arguments
+	if len(volID) == 0 {
+		return status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+
+	np.Lock.Lock()
+	defer np.Lock.Unlock()
+	if v, ok := np.VolumeMap[volID]; ok {
+		if _, err := os.Stat(v.targetPath); os.IsNotExist(err) {
+			os.Mkdir(v.targetPath, os.ModeDir)
+		}
```

Comment:
I don't believe you can publish to the same directory you stage to, so you'll need a different directory for publishing.

Additionally, you can remove the `os` import from this file. We don't need it right now, and it just creates side effects if you call this method in testing.

Reply:
Changed the published path location. Will add RPC for the methods in separate PR